lib/umbra/shadow_requester.rb

```diff
@@ -1,15 +1,19 @@
 module Umbra
   class ShadowRequester
+    attr_reader :queue, :count
+
     def initialize(count: 1, pool: 1, max_queue_size: 100)
       @count = count
-      @pool = pool
-      @max_queue_size = max_queue_size
+      @queue ||= Concurrent::CachedThreadPool.new(
+        min_threads: 1,
+        max_threads: pool,
+        max_queue: max_queue_size,
+        fallback_policy: :abort
+      )
     end
 
     def call(env)
-      @count.times do
-        queue << proc { RequestBuilder.call(env).run }
-      end
+      queue << proc { call!(env) }
```

Comment:
Wondering if instead of adding hydra as a dependency we could achieve the same thing by moving `RequestBuilder.call(env)` out of the loop, something like: 🤔 
```suggestion
      request = RequestBuilder.call(env)
      @count.times do
        queue << proc { request.run }
      end
```
That is main idea, right? Not sure if hydra provides something more or does something differently.
Maybe we need a way to benchmark these things? 😛

Reply:
Yes, I wanted to use `Hydra` in order to reduce the number of items in the `ShadowRequester` queue.

So allow `Hydra` (and therefore `Ethon` / `libcurl`) to handle concurrent requests instead of having multiple threads doing so by pulling off the queue.

The hypothesis is that this increases throughput of shadow requesting. By letting the parallelisation happen within `libcurl` instead of via `Concurrent::CachedThreadPool`.
I want to run a benchmark on production between the versions internally if possible before releasing a new version.

I'm not sure how `Typhoeus::Request` handles being called `#run` multiple times 🤔
I will have a look! 

related: [`Ethon::Multi`](https://rubydoc.info/github/typhoeus/ethon/Ethon%2FMulti:initialize)

Reply:
Updated so that the request is only built once, but queued multiple times - specs seems happy 😄